### PR TITLE
Make CreateDefaultClient() virtual in WebApplicationFactory<T>

### LINKED ISSUE #7635: Provide an ability to configure the HTTP client created by WebApplicationFactory<T>
I'm using `WebApplicationFactory<T>` for some tests where I need to configure `HttpClient` appropriately with some default headers in order for the application to accept requests as valid.

Because all the `HttpClient Create*Client()` methods are non-virtual, it's not possible for me to override any of the implementations to set the headers before returning the client, so I need to make my own method and get all my tests to call that instead (`CreateValidClient()`) rather than have the fixture handle it for the consumers of it transparently.

It would be much easier to use `WebApplicationFactory<T>` in these tests if some (if not all) of the methods for creating `HttpClient` instances were virtual.

This is similar to #7631 which also requests changes to make it easier to make derived classes from `WebApplicationFactory<T>`.

### PR DESCRIPTION
**Pull request by martincostello**
Make `CreateDefaultClient(Uri, DelegatingHandler[])` virtual to make it easier for consumers of `WebApplicationFactory<T>` to modify created instances of `HttpClient` in derived types.

Resolves #7635.

### CODE REVIEW

**Comment by javiercn** [comment]
I'm actually fine with making this method virtual as is the most basic building block for clients.

**Comment by mkArtakMSFT** [comment]
I don't feel super strong about either option. @rynowak, any thoughts?

File: src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationFactory.cs [hunk 1]
```diff
@@ -293,7 +293,7 @@ protected virtual void ConfigureWebHost(IWebHostBuilder builder)
         /// <param name="handlers">A list of <see cref="DelegatingHandler"/> instances to set up on the
         /// <see cref="HttpClient"/>.</param>
         /// <returns>The <see cref="HttpClient"/>.</returns>
-        public HttpClient CreateDefaultClient(Uri baseAddress, params DelegatingHandler[] handlers)
+        public virtual HttpClient CreateDefaultClient(Uri baseAddress, params DelegatingHandler[] handlers)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationFactory.cs]: I'd suggest  the following instead. 

Add a new method to `protected virtual CreateClientCore(params DelegatingHandler[] handlers)`

`HttpClient` has a lot of different things to configure. Whichever method is the **most fundamental** can't bake in things like `baseAddress`.

These convenience overloads should look like:
```
public HttpClient CreateDefaultClient(Uri baseAddress, params DelegatingHandler[] handlers)
{
    var client = CreateClientCore(handlers);
    client.BaseAddress = baseAddress;
    return client;
}
```

You could pivot back and forth on whether or not the most fundamental method is `Create` or a `Configure`.

This way someone can define additional overload

**Comment by martincostello** [comment]
Thanks for the feedback all. What’s the best approach for me to take for updating this PR?

I like the idea of a core creation method for the client/handler as a protected method that can be overridden, but at the same time the current method does things to private members that a replacement implementation would not have access to.

**Comment by martincostello** [comment]
What about something like this? This combines both approaches with virtual methods to override creating the root `HttpClient` and `HttpMessageHandler`, as well as a method for configuring the client before it is returned:

```csharp
public HttpClient CreateDefaultClient(Uri baseAddress, params DelegatingHandler[] handlers)
{
    EnsureServer();

    HttpClient client;

    if (handlers == null || handlers.Length == 0)
    {
        client = CreateClientCore();
    }
    else
    {
        for (var i = handlers.Length - 1; i > 0; i--)
        {
            handlers[i - 1].InnerHandler = handlers[i];
        }

        var serverHandler = CreateHandlerCore();
        handlers[handlers.Length - 1].InnerHandler = serverHandler;

        client = new HttpClient(handlers[0]);

        _clients.Add(client);
    }

    client.BaseAddress = baseAddress;

    ConfigureDefaultClient(client);

    return client;
}

protected virtual HttpClient CreateClientCore() => _server.CreateClient();

protected virtual HttpMessageHandler CreateHandlerCore() => _server.CreateHandler();

/// <summary>
/// Gives a fixture an opportunity to configure an HTTP client before it is returned.
/// </summary>
/// <param name="client">The <see cref="HttpClient"/> to configure.</param>
protected virtual void ConfigureDefaultClient(HttpClient client)
{
}
```

**Comment by javiercn** [comment]
We chatted about this, what we are doing is the following.
We are making `HttpClient CreateDefaultClient(params DelegatingHandler[] handlers)` virtual and having all other methods call into it.

**Comment by mkArtakMSFT** [comment]
Hi @martincostello. Will you be able to update your PR with @javiercn's suggestion today? We're locking down the changes going into the milestone today, so to take this change in we have just hours left.
If you'll not be able to update this PR by ~4PM PST, we'll make the change on our end to get this in 2.1 release.

**Comment by mkArtakMSFT** [comment]
No problem, @martincostello. we'll make the change then. Thanks for the effort anyway - it's valuable. I'm closing this PR in favor of the one we'll send out shortly.

**Pull request closed without merging**